src/server/pachyderm_test.go

```diff
@@ -9801,6 +9801,61 @@ func TestStandbyTransitions(t *testing.T) {
 	flushEventsAndCheckRC(initialRC.ResourceVersion)
 }
 
+func TestDatumSetCache(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping integration tests in short mode")
+	}
+	c := tu.GetPachClient(t)
+	c = c.WithDefaultTransformUser("1000")
+	require.NoError(t, c.DeleteAll())
+	dataRepo := tu.UniqueString("TestDatumSetCache_data")
+	require.NoError(t, c.CreateRepo(dataRepo))
+	masterCommit := client.NewCommit(dataRepo, "master", "")
+	require.NoError(t, c.WithModifyFileClient(masterCommit, func(mfc client.ModifyFile) error {
+		for i := 0; i < 90; i++ {
+			require.NoError(t, mfc.PutFile(strconv.Itoa(i), strings.NewReader("")))
+		}
+		return nil
+	}))
+	pipeline := tu.UniqueString("TestDatumSetCache")
+	_, err := c.PpsAPIClient.CreatePipeline(context.Background(),
+		&pps.CreatePipelineRequest{
+			Pipeline: client.NewPipeline(pipeline),
+			Transform: &pps.Transform{
+				Cmd: []string{"bash"},
+				Stdin: []string{
+					fmt.Sprintf("cp /pfs/%s/* /pfs/out/", dataRepo),
+					"sleep 1",
+				},
+			},
+			Input:        client.NewPFSInput(dataRepo, "/*"),
+			DatumSetSpec: &pps.DatumSetSpec{Number: 1},
+		})
+	require.NoError(t, err)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		ticker := time.NewTicker(time.Minute)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				tu.DeletePod(t, "etcd")
```

Comment:
The issue I was hitting was due to the master lock retry being pretty long ~20s in the test case, which made me think the job logic was stuck. When the master lock is lost, we hit a "fatal" error in the job logic, which will get retried at a higher level when the master lock is reacquired. Bumping up the test to be a bit longer and tweaking the delete pod frequency addresses the issue by still introducing failures fast enough to prevent full job completion, but slow enough to allow some room between the retry backoff and some processing.